### PR TITLE
Set Node Unhealthy in E2E by Stopping Kubelet

### PR DESCRIPTION
**Pull request by razo7**
The E2E test should emphasize a realistic scenario of only when a node is unhealthy FAR CR should be created. We achieve that by stopping kubelet on the desired node.

### CODE REVIEW

**Review by slintes — APPROVED** [review]

File: test/e2e/far_e2e_test.go [hunk 1]
```diff
@@ -201,17 +206,19 @@ func buildNodeParameters(clusterPlatformType configv1.PlatformType) (map[v1alpha
 	return testNodeParam, nil
 }
 
-// randomizeWorkerNode returns a worker node name which is different than the previous one
+// randomizeWorkerNode returns a worker node that his name is different than the previous one
```
Comment (mshitrit) [inline test/e2e/far_e2e_test.go:209]: IIUC this returns a random node out of the given node list.

>that his name is different than the previous one

I don't understand this part

Comment (razo7) [reply test/e2e/far_e2e_test.go:209]: Since we randomly pick a worker node and there are only 3 available nodes then we can pick the same node, even though we use a different seed for the random function. Therefore, we store the name of the last unhealthy node so we won't try to create second CR for the same node. 
> previous one

Refers to he previous remediation, so on the first remediation previousNodeName is nil, and on the second remediation it stores the name of the previous unheathy node in order to not create a CR on that node again.

Comment (mshitrit) [reply test/e2e/far_e2e_test.go:209]: I see now.
I advise against such implementation.
- IMO it's bad practice to use global var in such way (not concurrent safe and confusing)
- I think the method is confusing
- we have no idea how many iteration will occur before the for condition is broker because we can always get randomly the same node, also seems like a risk of a deadlock.

Instead I suggest the following , let `randomizeWorkerNode` do just that - get a random node out of the list:
```
func randomizeWorkerNode(nodes *corev1.NodeList) *corev1.Node {
	// Generate a random seed based on the current time
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	// Randomly select a worker node
	return &nodes.Items[r.Intn(len(nodes.Items))]
}
```

And just pass to it an already filtered list (filter out the node you don't want it to get).

File: test/e2e/utils/command.go [hunk 2]
```diff
@@ -24,10 +24,21 @@ import (
 	"github.com/medik8s/fence-agents-remediation/api/v1alpha1"
 )
 
+// StopKubelet runs cmd connand to stop kubelet for the node and returns an error only if it fails
```
Comment (clobrano) [inline test/e2e/utils/command.go:27]: typo s/connand/command

File: test/e2e/utils/command.go [hunk 3]
```diff
@@ -24,10 +24,21 @@ import (
...
+	cmd := "microdnf install util-linux -y && /usr/bin/nsenter -m/proc/1/ns/mnt /bin/systemctl stop kubelet"
+	_, err := runCommandInCluster(c, nodeName, testNsName, cmd, log)
+	if err != nil && strings.Contains(err.Error(), "connection refused") {
+		log.Info("ignoring expected error when stopping kubelet", "error", err.Error())
+		return nil
+	}
+	return err
+}
+
 // GetBootTime gets the boot time of the given node by running a pod on it executing uptime command
```
Comment (clobrano) [inline test/e2e/utils/command.go:38]: nit: function description should not state how the function works, just what it is expected to do (also, not part of this PR, so no need to change it here)

File: test/e2e/far_e2e_test.go [hunk 4]
```diff
@@ -99,12 +100,29 @@ var _ = Describe("FAR E2e", func() {
...
+		When("running FAR to reboot a node", func() {
+			It("should successfully remediate it", func() {
+				checkRemediation(nodeName, succeesRebootMessage, nodeBootTimeBefore)
+			})
+		})
+
 		When("running FAR to reboot two nodes", func() {
 			It("should successfully remediate the first node", func() {
 				checkRemediation(nodeName, succeesRebootMessage, nodeBootTimeBefore)
+				// run the second remediation on a different node
```
Comment (slintes) [inline test/e2e/far_e2e_test.go:119]: This looks like setup code for the next test, it shouldn't be here.
I think all this node selection stuff can and should be in the `BeforeEach()` block above...

**Review by slintes — CHANGES_REQUESTED** [review]

File: test/e2e/far_e2e_test.go (hunk 4 above)
Comment (razo7) [reply test/e2e/far_e2e_test.go]: I think it suits better an AfterEach, so I have used it.

File: test/e2e/far_e2e_test.go [hunk 5]
```diff
@@ -99,10 +100,29 @@ var _ = Describe("FAR E2e", func() {
...
 			DeferCleanup(deleteFAR, far)
 		})
+		When("running FAR to reboot one node", func() {
+			It("should successfully remediate it", func() {
+				checkRemediation(nodeName, nodeBootTimeBefore)
+			})
+		})
+
 		When("running FAR to reboot two nodes", func() {
+			AfterEach(func() {
```
Comment (slintes) [inline test/e2e/far_e2e_test.go:117]: I don't understand what you're trying to do here...

Can you please put this node selection stuff into the existing BeforeEach(), as suggested earlier already?

**Review by slintes — CHANGES_REQUESTED** [review]

File: test/e2e/far_e2e_test.go [hunk 6]
```diff
@@ -75,33 +73,53 @@ var _ = Describe("FAR E2e", func() {
...
 			nodeBootTimeBefore, err = e2eUtils.GetBootTime(clientSet, nodeName, testNsName, log)
 			Expect(err).ToNot(HaveOccurred(), "failed to get boot time of the node")
+
+			// set the node as "unhealthy" by disabling kubelet
+			makeNodeUnready(selectedNode)
+
 			far := createFAR(nodeName, fenceAgent, testShareParam, testNodeParam)
 			DeferCleanup(deleteFAR, far)
 		})
+		When("running FAR to reboot one node", func() {
```
Comment (slintes) [inline test/e2e/far_e2e_test.go:117]: Probably caused by rebasing etc., but I think this `When()` block can be removed. We test 2 remediations in the `When()` block below, that's enough I think :)

**Review by slintes — COMMENTED** [review]
one last thing... other than that lgtm

**Review by slintes — APPROVED** [review]
